### ISSUE TITLE
`valid-typeof` with `requireStringLiterals` should allow two `typeof` comparisons

### ISSUE THREAD

**Issue by ljharb** [eslint/eslint#7078]
**What version of ESLint are you using?**
v3.4.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
just `'valid-typeof': ['error', { requireStringLiterals: true }]`

**What did you do? Please include the actual source code causing the issue.**
`echo 'typeof foo === typeof bar' | eslint --stdin --no-eslintrc --rule valid-typeof:'[2,{requireStringLiterals: true}]'`

**What did you expect to happen?**
No error.

**What actually happened? Please include the actual, raw output from ESLint.**
```sh
text>
  1:1   error  Typeof comparisons should be to string literals  valid-typeof
  1:16  error  Typeof comparisons should be to string literals  valid-typeof

✖ 2 problems (2 errors, 0 warnings)
```

**Comment by platinumazure** [comment]
Cripes. On the one hand the option name seems fairly explicit, on the other hand I can see why this is a pain.

**Comment by vitorbal** [comment]
Judging by the examples in the docs, this was intentional, but looking back at #6698, it might have been an oversight.
@not-an-aardvark @btmills tagging you two as you were champion/implementer of the option. Any thoughts? Should this be considered a bug?

**Comment by not-an-aardvark** [comment]
At the moment, the option explicitly only allows comparisons to string *literals*, and it was implemented with that in mind (`typeof a === typeof b` is reported as an error in the docs and the tests).

However, I think the original goal of the option in #6698 was to be able to statically prove that a `typeof` comparison is valid, regardless of whether it actually uses string literals.

The regular `valid-typeof` rule doesn't accomplish this, since the following is not reported:

```js
const foo = 'not a valid typeof string';
typeof myValue === foo; // invalid, but not reported
```

On the other hand, the `requireStringLiterals` option is too strict for this goal, since `typeof a === typeof b` is always valid, but is reported as an error.

I don't think changing the behavior of the `requireStringLiterals` option is a good idea, since the name `requireStringLiterals` clearly implies that string literals should be required. Maybe we should add a new option (`strictValidation`?) to allow `typeof a === typeof b`, and deprecate `requireStringLiterals`.

**Comment by ljharb** [comment]
The spirit behind "requireStringLiterals" imo is as you said - to statically prove that a comparison is valid. Comparing to string literals *or* to another `typeof` comparison achieves that.

Perhaps the option name isn't correct, but the behavior imo should be that `typeof a === typeof b` is always considered valid.

**Comment by btmills** [comment]
How bad would it be if `requireStringLiterals` permitted `typeof a === typeof b`, adding the exception without changing the name? That's the easiest route. It's compatible with the spirit of the option albeit not the name.

**Comment by platinumazure** [comment]
I'm going to do one better (slash crazier): Given that we haven't released,
can we fix behavior and name before tomorrow's minor release?

**Comment by btmills** [comment]
> Given that we haven't released

The new option was included in [3.4.0](http://eslint.org/blog/2016/08/eslint-v3.4.0-released#enhancements). If it hadn't been released yet, I'd actually like that idea 😈

**Comment by platinumazure** [comment]
Oh damnit, I thought we had just put it in this week. Sorry, I'm tracking (trying to track) too many PRs.

**Comment by ljharb** [comment]
https://github.com/eslint/eslint/issues/7078#issuecomment-245669335 Is exactly what I'm suggesting, and it'd be semver-patch.

**Comment by platinumazure** [comment]
I can live with that for now. Maybe in a major (5.0?) we can rename the option or something.

**Comment by not-an-aardvark** [comment]
Working on this.
